### PR TITLE
`find*` methods include navigational properties

### PR DESCRIPTION
**Pull request by nabdelgadir**
Resolves https://github.com/strongloop/loopback-next/issues/2632.

- `find` and `findById` now use `T & Relations` instead of `T`
- New prompt for `lb4 model` that asks for relations in order to create the relation interface (e.g. `export interface TodoRelations {}`). It leaves it empty if there are no current relations.
- Newly scaffolded models from `lb4 model` include `{modelName}WithRelations` type (e.g. `export type TodoWithRelations = Todo & TodoRelations;`)
- Generic parameter `Relations` added to `DefaultCrudRepository`
- When scaffolding a repository with `lb4 repository`, it now includes `{modelName}Relations` (e.g. 
```ts
export class TodoListRepository extends DefaultCrudRepository<
  TodoList,
  typeof TodoList.prototype.id
  TodoListRelations
> 
```
)

**Question**: should all models include a relations interface? If so, I'll update the examples/docs.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: packages/cli/generators/model/index.js [hunk 1]
```diff
@@ -297,6 +298,52 @@ module.exports = class ModelGenerator extends ArtifactGenerator {
...
+      return this.exit(err);
+    }
+
+    // skip if there are no other models
+    if (!models.length) return;
+
+    return this.prompt([
+      {
+        name: 'relations',
+        message: 'Is this model part of any relations?',
```
Comment (bajtos) [inline packages/cli/generators/model/index.js]: Very recently, we landed `lb4 relation` command - see https://github.com/strongloop/loopback-next/commit/75939a4a144b3068ff2890394c178faad58d7458.

I find it weird to have `lb4 model` prompting for model relations, when there is a dedicated command that does better job at scaffolding relation-based code.

Is this part addressing the following acceptance criteria?

> Update lb4 model to start emitting `{ModelName}Relations` interfaces

As I was envisioning the solution, an empty interface should be emitted for all newly-created models, so that we can _Update `lb4 repository` to include `{ModelName}Relations` in the generic parameters of `DefaultCrudRepository` we are inheriting from_

Entries for individual relations would be added to `{ModelName}Relations` either manually by the application developer or by `lb4 repository` command.

My conclusions:
- The answer to your _Question: should all models include a relations interface?_ is YES.
- These new `lb4 model` prompts should be removed.
- Ideally, `lb4 repository` should be improved to add entries to relations interfaces. This was not part of the original acceptance criteria and I don't know how much effort would be required, it may be best to leave this part out of scope of this pull request and possibly out of scope of #2960 too.

What do you think?

Comment (nabdelgadir) [reply packages/cli/generators/model/index.js]: @bajtos thanks for the comment!

> Is this part addressing the following acceptance criteria?
> 
> > Update `lb4 model` to start emitting `{ModelName}Relations` interfaces

Yes, initially I thought calling `lb4 model` should also populate the interface, but I agree that it should be removed and left up to another CLI command.

> The answer to your Question: should all models include a relations interface? is YES.
> These new `lb4 model` prompts should be removed.

👍 

> Ideally, `lb4 repository` should be improved to add entries to relations interfaces. This was not part of the original acceptance criteria and I don't know how much effort would be required, it may be best to leave this part out of scope of this pull request and possibly out of scope of #2960 too.

I originally thought it was, but I misunderstood the acceptance criteria, but I'm good with opening a new issue for this as it's out of scope of this PR.

Comment (nabdelgadir) [reply packages/cli/generators/model/index.js]: Issue created for adding entries to the interfaces: https://github.com/strongloop/loopback-next/issues/2988

In the new commits, I added empty interfaces because I think they should be populated only after https://github.com/strongloop/loopback-next/issues/2633 is landed. What do you think?

File: examples/todo-list/src/models/todo-list-image.model.ts [hunk 2]
```diff
@@ -29,3 +29,7 @@ export class TodoListImage extends Entity {
     super(data);
   }
 }
+
+export interface TodoListImageRelations {}
```
Comment (bajtos) [inline examples/todo-list/src/models/todo-list-image.model.ts]: `TodoListImage` already have some relations defined (I think it belongs to TodoList?), can you add navigational properties for these relations to this interface please?

File: examples/todo-list/src/models/todo-list.model.ts [hunk 3]
```diff
@@ -36,3 +36,7 @@ export class TodoList extends Entity {
     super(data);
   }
 }
+
+export interface TodoListRelations {}
```
Comment (bajtos) [inline examples/todo-list/src/models/todo-list.model.ts]: `TodoList` already have some relations defined (I think it has many `TodoListImage` and `Todo` instances), can you add navigational properties for these relations to this interface please?

File: examples/todo-list/src/models/todo.model.ts [hunk 4]
```diff
@@ -41,3 +41,7 @@ export class Todo extends Entity {
     super(data);
   }
 }
+
+export interface TodoRelations {}
```
Comment (bajtos) [inline examples/todo-list/src/models/todo.model.ts]: `Todo` already have some relations defined (I think it belongs to TodoList?), can you add navigational properties for these relations to this interface please?

File: examples/todo/src/models/todo.model.ts [hunk 5]
```diff
@@ -44,3 +44,7 @@ export class Todo extends Entity {
     super(data);
   }
 }
+
+export interface TodoRelations {}
```
Comment (bajtos) [inline examples/todo/src/models/todo.model.ts]: In this example, `Todo` does not have any relations defined, thus `TodoRelations` should be an empty interface :+1:

To reduce noise in `git diff` output when adding new relations in the future, I am proposing to add a line comment into the empty object block as follows.

```ts
export interface TodoRelations {
  // describe navigational properties here
}
```

This should be applied to all places where we have an empty Relations interface and to the CLI template too.

Thoughts?

File: packages/cli/generators/model/index.js [hunk 6]
```diff
@@ -297,12 +296,6 @@ module.exports = class ModelGenerator extends ArtifactGenerator {
         if (this.artifactInfo.allowAdditionalProperties) {
           Object.assign(this.artifactInfo.modelSettings, {strict: false});
         }
-
-        this.log(
-          `Let's add a property to ${chalk.yellow(
-            this.artifactInfo.className,
```
Comment (bajtos) [inline packages/cli/generators/model/index.js]: This change seems unrelated to the rest of this pull request, what is the reasoning for removing this message?

File: packages/cli/generators/model/templates/model.ts.ejs [hunk 7]
```diff
@@ -33,3 +33,9 @@ export class <%= className %> extends <%= modelBaseClass %> {
     super(data);
   }
 }
+
+export interface <%= className %>Relations {
+
```
Comment (bajtos) [inline packages/cli/generators/model/templates/model.ts.ejs]: As I commented above, let's replace this empty line (that will be removed by prettier) with a line comment.

File: packages/repository/src/__tests__/fixtures/models/address.model.ts [hunk 8]
```diff
@@ -29,3 +29,7 @@ export class Address extends Entity {
   @belongsTo(() => Customer)
   customerId: number;
 }
+
+export interface AddressRelations {}
```
Comment (bajtos) [inline packages/repository/src/__tests__/fixtures/models/address.model.ts]: Please add navigational properties for existing relations of `Address` model.

File: packages/repository/src/__tests__/fixtures/models/order.model.ts [hunk 9]
```diff
@@ -29,3 +29,7 @@ export class Order extends Entity {
   @belongsTo(() => Customer)
   customerId: number;
 }
+
+export interface OrderRelations {}
```
Comment (bajtos) [inline packages/repository/src/__tests__/fixtures/models/order.model.ts]: Please add navigational properties for existing relations of `Customer` model.

File: packages/cli/generators/model/index.js (hunk 6 above)
Comment (nabdelgadir) [reply packages/cli/generators/model/index.js]: Removed by accident, will add back 👍

File: examples/todo/src/models/todo.model.ts (hunk 5 above)
Comment (nabdelgadir) [reply examples/todo/src/models/todo.model.ts]: I agree 👍

**Comment by nabdelgadir** [comment]
@bajtos 
> IMO, when a model already has some relations defined, then the `*Relations` interface should have the navigational properties defined too. We will need these navigational properties for our next steps on the journey to implement the feature "inclusion of related models".
>
> As I see it, #2633 is about automating the task of adding navigational properties. Until it's done, developers will have to define these properties manually, which is exactly what I am asking you to do in this pull request too.

I thought the purpose of #2633 was only to add them to the `TodoList` example and verify it works for that example (and then after it's tested there, then we could add the properties of the `*Relations` interface to the other examples/docs). But I guess adding them here makes more sense, so now they're a part of this PR 👍 

Also, did you mean https://github.com/strongloop/loopback-next/issues/2988 was for automating instead of #2633?

**Comment by bajtos** [comment]
> Also, did you mean #2988 was for automating instead of #2633?

Yes, I meant #2988. Sorry for the confusion!

File: docs/site/BelongsTo-relation.md [hunk 10]
```diff
@@ -212,14 +223,21 @@ export class Category extends Entity {
     super(data);
   }
 }
+
+export interface CategoryRelations {
+  categories?: CategoryWithRelations[];
+}
```
Comment (bajtos) [inline docs/site/BelongsTo-relation.md]: I think this interface should include a navigation property for `parent` relation too, WDYT?

```ts
export interface CategoryRelations {
  categories?: CategoryWithRelations[];
  parent?: CategoryWithRelations;
}
```
